src/hotspot/share/gc/parallel/psCardTable.cpp

```diff
@@ -277,7 +289,15 @@ void PSCardTable::scavenge_contents_parallel(ObjectStartArray* start_array,
     {
       // Identify right ends.
       HeapWord* obj_addr = start_array->object_start(cur_stripe_end_addr - 1);
-      HeapWord* obj_end_addr = obj_addr + cast_to_oop(obj_addr)->size();
+      size_t obj_sz = cast_to_oop(obj_addr)->size();
+      HeapWord* obj_end_addr = obj_addr + obj_sz;
+      // large arrays starting in this stripe are scanned here
+      if (obj_sz >= large_obj_arr_min_words() && cast_to_oop(obj_addr)->is_objArray() &&
+          obj_addr >= cur_stripe_addr) {
+        // the last condition is not redundant as we can reach here if an obj starts at space_top
```

Comment:
I think the comment here must be expanded. As far as I understand you, one can generally assume that the array found here must start in the stripe because either:
- the array actually starts in the stripe. Since it is larger than a stripe it must be the rightmost object.
- the array covers the whole stripe and doesn't start in the stripe. In this case we would never land here since the check !start_array->object_starts_in_range(cur_stripe_addr, cur_stripe_end_addr) in line 258 filters this out already (no object would start in the stripe)
- the array ends in the stripe, but not on the stripe end. In this case the rightmost object is not the array itself..

But when the object start array overlaps with the space_top, the start_array->object_starts_in_range(cur_stripe_addr, cur_stripe_end_addr) could return true, if there is an object at space_top. In this case the array would not start in the stripe and the check ensures that.

Reply:
Thanks for looking at the PR!
I agree that it's not easy to see what's intended. And it's not that easy to explain either.
It might be possible to remove the condition and the comment if we check `first_obj_addr >= cur_stripe_end_addr` in L279 and just continue if true.

Reply:
The guiding principle of `PSCardTable::scavenge_contents_parallel` is to scan only objects that start in the current stripe. There is a special case for the last stripe ending at `space_top` if `space_top` is not card-aligned and an object starts there. With https://github.com/openjdk/jdk/pull/14846/commits/0eb924e1d4dfa1f751a55038063af4f6092be8c2 this is handled explicitly and more clearly to make sure stripes without object starts are skipped right away. This is important for correct handling of large object arrays. IMHO also helpful aside from large arrays.

Reply:
> The guiding principle of PSCardTable::scavenge_contents_parallel is to scan only objects that start in the current stripe

And with this pr also large array chunks. When we found that the last stripe before `space_top` has no object start then we know there is no array chunk to scan either. There might be the end of a large array but this is scanned together with the chunk in the previous stripe (see https://github.com/openjdk/jdk/pull/14846/files#diff-fc75bbdc27c5b981a8b66f77cbd83e6d86d8f7d3e313701207d880b7036c8b92R238-R239).

Reply:
I didn't like the last commit (https://github.com/openjdk/jdk/commit/0eb924e1d4dfa1f751a55038063af4f6092be8c2): it introduced a condition that is only needed for large array handling adding another hunk to the patch.

Instead I've expanded the constraints

```c++
    // 4. range of large array elements to be scanned: [first_obj_addr, cur_stripe_end_addr)
    //    limited to dirty regions
```

It is enforced where large arrays are handled. Comments explain the assumptions and assertions check them.